tornettools/parse_tgen.py

```diff
@@ -198,16 +204,46 @@ def __get_client_goodput(data, startts, stopts):
             db = data['data'][name]
             streams = db['tgen']['streams']
 
+
             for sid in streams:
                 stream = streams[sid]
-                if "elapsed_seconds" in stream and \
-                    "payload_bytes_recv" in stream['elapsed_seconds']:
-                    bytes_db = stream['elapsed_seconds']['payload_bytes_recv']
-                    if '512000' in bytes_db and '1048576' in bytes_db:
-                        seconds = float(bytes_db['1048576']) - float(bytes_db['512000'])
-                        bytes = 1048576 - 512000
-                        mbit = bytes/1048576.0*8.0
-                        mbit_per_second = mbit/seconds
-                        goodput.append(mbit_per_second)
+                if "elapsed_seconds" in stream:
+                    if download_size == "1 MiB":
+                        recvsize = "1048576"
+                        es_1MiB = stream['elapsed_seconds']['payload_bytes_recv']
+                        goodput = __get_client_goodput_1MiB(es_1MiB, recvsize, goodput)
+                    elif download_size == "5 MiB":
+                        recvsize = stream["stream_info"]["recvsize"]
+                        es_5MiB = stream["elapsed_seconds"]["payload_progress_recv"]
+                        goodput = __get_client_goodput_5MiB(es_5MiB, recvsize, goodput)
+                    else:
+                        continue
+
+    return goodput
 
+def __get_client_goodput_1MiB(es, recvsize, goodput):
+    # Tor computs goodput based on the time between the .5 MiB byte to the 1 MiB byte.
+    # Ie to cut out circuit build and other startup costs.
+    # https://metrics.torproject.org/reproducible-metrics.html#performance
+    if recvsize in es and '512000' in es:
+        start, end = float(es['512000']), float(es['1048576'])
+        mbit_per_second = __goodput_mbps(start, end, 512000, 1048576)
+        if mbit_per_second:
+            goodput.append(mbit_per_second)
     return goodput
+
+def __get_client_goodput_5MiB(es, recvsize, goodput):
+    # For 5 MiB downloads we extract the time (in seconds) elapsed between
+    # receiving the 4MiB byte and 5MiB byte, which is a total amount of 1 MiB  or
+    # 8 Mib.
+    if recvsize == "5242880" and "1.0" in es:
+        start, end = float(es["0.8"]), float(es["1.0"])
+        mbit_per_second = __goodput_mbps(start, end, 4194304, 5242880)
+        if mbit_per_second:
+            goodput.append(mbit_per_second)
+    return goodput
+
+def __goodput_mbps(start_time, end_time, start_bytes, end_bytes):
+    if start_time > 0 and end_time > 0 and end_time > start_time:
+        return ((end_bytes - start_bytes) / 1000000.0) * 8.0  / (end_time - start_time)
```

Comment:
Yes I changed it to make the computation uniform. In one case (the onionperf calculated throughput from the live data) we divide by 1000000.0 during plot, and in the tgen results we divide by 2^20 when we calculate it.

Reply:
Again I agree with the old data that would have to be parsed again. At the same time it would be nice if the throughput could be computed with the same calculation.